mayastor/src/subsys/mbus/mod.rs

```diff
@@ -0,0 +1,113 @@
+//! Message bus connecting mayastor to the control plane components.
+//!
+//! It is designed to make sending events to control plane easy in the future.
+//!
+//! A Registration subsystem is used to keep moac in the loop
+//! about the lifecycle of mayastor instances.
+
+pub mod mbus_nats;
+pub mod registration;
+
+use crate::core::{MayastorCliArgs, MayastorEnvironment};
+use async_trait::async_trait;
+use mbus_nats::NATS_MSG_BUS;
+use registration::Registration;
+use smol::io;
+use spdk_sys::{
+    spdk_add_subsystem,
+    spdk_subsystem,
+    spdk_subsystem_fini_next,
+    spdk_subsystem_init_next,
+};
+use structopt::StructOpt;
+
+// wrapper around our MBUS subsystem used for registration
+pub struct MessageBusSubsystem(*mut spdk_subsystem);
+
+impl Default for MessageBusSubsystem {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl MessageBusSubsystem {
+    /// initialise a new subsystem that handles the control plane
+    /// message bus registration process
+    extern "C" fn init() {
+        debug!("mayastor mbus subsystem init");
+        let args = MayastorEnvironment::new(MayastorCliArgs::from_args());
+        if let (Some(_), Some(grpc)) = (args.mbus_endpoint, args.grpc_endpoint)
+        {
+            Registration::init(&args.name, &grpc);
+        }
+        unsafe { spdk_subsystem_init_next(0) }
+    }
+
+    extern "C" fn fini() {
+        debug!("mayastor mbus subsystem fini");
+        let args = MayastorEnvironment::new(MayastorCliArgs::from_args());
+        if args.mbus_endpoint.is_some() && args.grpc_endpoint.is_some() {
+            Registration::get().fini();
+        }
+        unsafe { spdk_subsystem_fini_next() }
+    }
+
+    fn new() -> Self {
+        info!("creating Mayastor mbus subsystem...");
+        let args = MayastorEnvironment::new(MayastorCliArgs::from_args());
+        if let Some(nats) = args.mbus_endpoint {
+            mbus_nats::message_bus_init(nats);
+        }
```

Comment:
I don't think we have to call this here.  The init/fini is to ensure you get called before we start up fully, but it is not needed to start the bus here. The reason I say this that I'm a bit concerned about what creating the runtime in the init call does.

Having that said, this runtime mess is not something of our doing, but.. I do have an idea to make it less horrible but... i might get to that later...

Reply:
It's was here for convenience only. If you remember the registration "run" was being spawned in the init call so I tried to make sure that nats was already ready then by having it connect in the "new" as it is now. But we now run the registration in the "env" tokio loop so we can indeed move the message bus init elsewhere...